src/charm.py

```diff
@@ -170,6 +176,64 @@ def _configure_charm_authorization(self):
             login_details_secret = self.model.get_secret(label=GOCERT_LOGIN_SECRET_LABEL)
             login_details_secret.set_content(login_details.to_dict())
 
+    def _configure_certificate_requirers(self):
+        """Get all CSR's and certs from databags and Gocert, compare differences and update requirers if needed."""
+        login_details = self._get_or_create_admin_account()
+        if not login_details or not login_details.token:
+            logger.warning("couldn't distribute certificates: not logged in")
+            return
+        databag_csrs = self.tls.get_certificate_requests()
+        gocert_csrs_table = self.client.get_certificate_requests_table(login_details.token)
+        if not gocert_csrs_table:
+            logger.warning("couldn't distribute certificates: couldn't get table from GoCert")
+            return
+
+        for request in databag_csrs:
+            # Find all rows in the table that match the request's CSR
+            matching_rows = list(
+                filter(
+                    lambda x: x.csr == str(request.certificate_signing_request),
+                    gocert_csrs_table.rows,
+                )
+            )
+            # If there are no matches, upload CSR to GoCert
+            if len(matching_rows) < 1:
+                self.client.post_csr(str(request.certificate_signing_request), login_details.token)
+                continue
+            matching_row = matching_rows[0]
+            matching_provided_certificates = list(
+                filter(
+                    lambda x: str(x.certificate_signing_request) == matching_row.csr,
+                    self.tls.get_issued_certificates(request.relation_id),
+                )
+            )
+            if (
+                matching_row.certificate_chain == "rejected"
+                or matching_row.certificate_chain == ""
+            ):
+                # TODO: tls lib v4 lacks the ability to remove the certificate of a specific request
+                continue
+            certificate_chain = [
+                Certificate.from_string(cert) for cert in matching_row.certificate_chain
+            ]
+            if (
+                len(certificate_chain) > 0
```

Comment:
Why is it important to check the length of the chain here? Additionally we might enter the next condition check with an empty chain which could cause an index out of bound error, or am I reading this wrong?

Reply:
It is not possible to be out of bounds because the length of the certificate chain can never be an empty list, it will either be a list of 2 items or more, the string "", or the string "rejected", and the second and third options will automatically continue the loop. This is a separate "bug" since (apparently) there's no way to remove a certificate from the relation, if the csr becomes rejected after a certificate was provided the stale cert will remain in the databag.

However, in the impossible chance that it may actually happen by someone modifying the serialize code in gocert, I added the check to both sides of the `or`.

This if statement is basically if `there's a certificate available but we didn't send it yet` or `we already sent a certificate but it's not the same one that's in gocert`

I cleaned it up a bit.